cli/compiler.rs

```diff
@@ -74,229 +96,552 @@ fn req(
+    let js_key = self
+      .disk_cache
+      .get_cache_filename_with_extension(module_specifier.as_url(), "js");
+    self
+      .disk_cache
+      .set(&js_key, contents.as_bytes())
+      .and_then(|_| {
+        self.mark_compiled(&module_specifier.to_string());
+
+        let source_file = self
+          .deno_dir
+          .fetch_source_file(&module_specifier)
+          .expect("Source file not found");
+
+        let version_hash = source_code_version_hash(
+          &source_file.source_code,
+          version::DENO,
+          &self.config_hash,
+        );
+
+        let compiled_file_metadata = CompiledFileMetadata {
+          source_path: source_file.filename.to_str().unwrap().to_string(),
+          version_hash,
+        };
+        let meta_key = self
+          .disk_cache
+          .get_cache_filename_with_extension(module_specifier.as_url(), "meta");
+        self.disk_cache.set(
+          &meta_key,
+          compiled_file_metadata.to_json_string()?.as_bytes(),
+        )
       })
-    }).and_then(move |module_meta_data_after_compile| {
-      // Explicit drop to keep reference alive until future completes.
-      drop(compiling_job);
-
-      Ok(module_meta_data_after_compile)
-    }).then(move |r| {
-      // TODO(ry) do this in worker's destructor.
-      // resource.close();
-      r
-    })
+  }
+
+  // TODO: ideally we shouldn't construct SourceFile by hand, but it should be delegated to
+  // SourceFileFetcher
+  pub fn get_source_map_file(
+    self: &Self,
+    module_specifier: &ModuleSpecifier,
+  ) -> Result<SourceFile, ErrBox> {
+    let cache_key = self
+      .disk_cache
+      .get_cache_filename_with_extension(module_specifier.as_url(), "js.map");
+    let source_code = self.disk_cache.get(&cache_key)?;
+    let source_map_filename = self.disk_cache.location.join(cache_key);
+    debug!("source map filename: {:?}", source_map_filename);
+
+    let source_map_file = SourceFile {
+      url: module_specifier.as_url().to_owned(),
+      redirect_source_url: None,
+      filename: source_map_filename,
+      media_type: msg::MediaType::JavaScript,
+      source_code,
+    };
+
+    Ok(source_map_file)
+  }
+
+  fn cache_source_map(
+    self: &Self,
+    module_specifier: &ModuleSpecifier,
+    contents: &str,
+  ) -> std::io::Result<()> {
+    let source_map_key = self
+      .disk_cache
+      .get_cache_filename_with_extension(module_specifier.as_url(), "js.map");
+    self.disk_cache.set(&source_map_key, contents.as_bytes())
+  }
+
+  pub fn cache_compiler_output(
+    self: &Self,
+    module_specifier: &ModuleSpecifier,
+    extension: &str,
+    contents: &str,
+  ) -> std::io::Result<()> {
+    match extension {
+      ".map" => self.cache_source_map(module_specifier, contents),
+      ".js" => self.cache_compiled_file(module_specifier, contents),
+      _ => unreachable!(),
+    }
+  }
+
+  fn try_to_resolve(self: &Self, script_name: &str) -> Option<ModuleSpecifier> {
```

Comment:
Move `try_to_resolve`,`try_resolve_and_get_source_file` and `try_to_resolve_and_get_source_map` to separate `impl TsCompiler` block